pkg/reduce/readloop/readloop.go

```diff
@@ -115,87 +115,156 @@ func (rl *ReadLoop) Startup(ctx context.Context) error {
 	return nil
 }
 
-// Process is one iteration of the read loop.
+// Process is one iteration of the read loop which writes the messages to the PBQs followed by acking the messages, and
+// then closing the windows that can closed.
 func (rl *ReadLoop) Process(ctx context.Context, messages []*isb.ReadMessage) {
-	// There is no Cap on backoff because setting a Cap will result in
-	// backoff stopped once the duration exceeds the Cap
+	// write messages to windows based by PBQs.
+	successfullyWrittenMessages, err := rl.writeMessagesToWindows(ctx, messages)
```

Comment:
To guarantee acks, we need to handle the context Done event here in addition to write

Reply:
why do we need to do that?

Reply:
We check for `ctx.Done()` befeore every batch iteration in https://github.com/numaproj/numaflow/blob/a19f77c017ddd706d4e6c2e15d457e34757c604a/pkg/reduce/reduce.go#L89-L91

Reply:
i guess thats fine. since we are cancelling it after we close the read, this should work.